americangut/question/ag_question.py

```diff
@@ -0,0 +1,117 @@
+        if not isinstance(description, str):
+            raise TypeError('description must be a string')
+        if not inspect.isclass(dtype):
+            raise TypeError('dtype must be a class')
+        if not isinstance(clean_name, str) and clean_name is not None:
+            raise TypeError('If supplied, clean_name must be a string')
+
+        # Handles the main information about the data
+        self.name = name
+        self.description = description
+        self.dtype = dtype
+
+        self.type = 'Question'
+        if clean_name is None:
+            self.clean_name = name.replace('_', ' ').title()
+        else:
+            self.clean_name = clean_name
+
+        # Sets up
+        self.free_response = kwargs.get('free_response', False)
+        self.mimmarks = kwargs.get('mimmarks', False)
+        self.ontology = kwargs.get('ontology', None)
+        self.remap_ = kwargs.get('remap', None)
+
+    def check_map(self, map_):
+        """Checks the group exists in the metadata
+
+        Parameters
+        ----------
+        map_ : DataFrame
+            A pandas object containing the data to be analyzed. The
+            Question `name` should be a column in the `map_`.
+
+        """
+        if self.name not in map_.columns:
+            raise ValueError('%s is not a column in the supplied map!'
+                             % self.name)
+
+    def remap_data_type(self, map_, watch=True):
+        """Makes sure the target column in map_ has the correct datatype
+
+        map_ : DataFrame
+            A pandas dataframe containing the column described by the question
+            name.
+        watch : bool
+            A flag to indicate the change should be logged. Should generally
+            be true, unless `remap_data_type` is called before another
+            function is executed.
+
+        """
+        if self.dtype == bool:
+            if not (set(map_[self.name].dropna()).issubset(
+                    self.true_values.union(self.false_values))):
+                raise TypeError('%s cannot be cast to a bool value.'
+                                % self.name)
+
+            def remap_(x):
+                if isinstance(x, str) and x.lower() in self.true_values:
+                    return True
+                elif isinstance(x, str) and x.lower() in self.false_values:
+                    return False
+                elif np.isnan(x):
+                    return x
+                else:
+                    try:
+                        return bool(x)
+                    except:
+                        return np.nan
+        else:
+            def remap_(x):
+                return self.dtype(x)
+
+        map_[self.name] = map_[self.name].apply(remap_).astype(self.dtype)
+        map_.replace('nan', np.nan, inplace=True)
+
+        if watch and self.type in {'Categorical', 'Multiple', 'Clinical',
+                                   'Bool', 'Frequency'}:
+            self._update_order(remap_)
```

Comment:
Let me get this straight:

- an `AgQuestion` has a type, i.e., categorical
- an `AgQuestion` has a dtype, i.e., float
- an `AgQuestion` corresponds to an entry in the data dictionary

Shouldn't it be immutable since the data dictionary is immutable? Why would
a categories dtype change? Is there functionality that a derived class has
that doesn't make sense for all derived classes? What does "updating the
parent object" mean -- is this a singleton pattern? Are derivative types
created dynamically, if so, under what circumstances since the data
dictionary is static?

I'm sorry if I'm just totally lost here. This feels overly complex, but I
also don't fully understand what problem it is solving so it's hard to
evaluate. Can another developer review please?

On Wed, Jan 27, 2016 at 12:51 AM, J W Debelius <notifications@github.com>
wrote:

> In americangut/question/ag_question.py
> <https://github.com/biocore/American-Gut/pull/188#discussion_r50921561>:
>
> > +                    return x
> > +                else:
> > +                    try:
> > +                        return bool(x)
> > +                    except:
> > +                        return np.nan
> > +        else:
> > +            def remap_(x):
> > +                return self.dtype(x)
> > +
> > +        map_[self.name] = map_[self.name].apply(remap_).astype(self.dtype)
> > +        map_.replace('nan', np.nan, inplace=True)
> > +
> > +        if watch and self.type in {'Categorical', 'Multiple', 'Clinical',
> > +                                   'Bool', 'Frequency'}:
> > +            self._update_order(remap_)
>
> The base class can be instantiated. It's just really limited, and probably
> won't appear that way in the actual data dictionary function where these
> are being used.
>
> I'm struggling with this, because there aren't great, clean ways to handle
> the logic.
>
>    1. Break the method apart, and treat continuous and categorical data
>    differently. This solves the inheritance problem, but code gets replicated.
>    2. Move the _update_order function into AgQuestion. Give it the logic
>    to update_order depending on the type. This means updating the parent
>    object every time you need to create a new derivative type, if you want it
>    to behave correctly. That doesn't seem like the right solution.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/American-Gut/pull/188/files#r50921561>.
>